File: frontend/src/services/productosService.ts
import { apiRequest } from './authService';
import type {
  Producto,
  ProductoFormData,
  ProductoFilters,
  Categoria,
  ProductoImagen,
  ProductoVariante,
  PaginatedResponse,
  ApiResponse,
} from '@/types';

// Mappers para convertir entre formatos del frontend y backend
const toDTO = (data: ProductoFormData) => ({
  nombre: data.nombre,
  descripcion: data.descripcion,
  descripcion_corta: data.descripcion_corta,
  categoria: data.categoria_id,
  precio: data.precio,
  en_oferta: data.en_oferta || false,
  precio_oferta: data.precio_oferta,
  descuento_porcentaje: data.descuento_porcentaje || 0,
  fecha_inicio_oferta: data.fecha_inicio_oferta?.toISOString(),
  fecha_fin_oferta: data.fecha_fin_oferta?.toISOString(),
  stock: data.stock,
  stock_minimo: data.stock_minimo || 5,
  meses_garantia: data.meses_garantia || 12,
  descripcion_garantia: data.descripcion_garantia,
  sku: data.sku,
  codigo_barras: data.codigo_barras,
  marca: data.marca,
  modelo: data.modelo,
  peso: data.peso,
  activo: data.activo ?? true,
  destacado: data.destacado ?? false,
});

const fromDTO = (data: any): Producto => ({
  id: data.id,
  nombre: data.nombre,
  slug: data.slug,
  descripcion: data.descripcion,
  descripcion_corta: data.descripcion_corta,
  imagen: data.imagen,
  categoria: data.categoria,
  precio: parseFloat(data.precio),
  en_oferta: data.en_oferta,
  precio_oferta: data.precio_oferta ? parseFloat(data.precio_oferta) : null,
  precio_final: parseFloat(data.precio_final),
  descuento_porcentaje: data.descuento_porcentaje,
  fecha_inicio_oferta: data.fecha_inicio_oferta,
  fecha_fin_oferta: data.fecha_fin_oferta,
  stock: data.stock,
  stock_minimo: data.stock_minimo,
  meses_garantia: data.meses_garantia,
  descripcion_garantia: data.descripcion_garantia,
  sku: data.sku,
  codigo_barras: data.codigo_barras,
  marca: data.marca,
  modelo: data.modelo,
  peso: data.peso ? parseFloat(data.peso) : null,
  activo: data.activo,
  destacado: data.destacado,
  vistas: data.vistas,
  ventas: data.ventas,
  imagenes: data.imagenes || [],
  variantes: data.variantes || [],
  creado: data.creado,
  actualizado: data.actualizado,
});

const categoriaFromDTO = (data: any): Categoria => ({
  id: data.id,
  nombre: data.nombre,
  slug: data.slug,
  descripcion: data.descripcion,
  imagen: data.imagen,
  activa: data.activa,
  orden: data.orden,
  productos_count: data.productos_count || 0,
  creado: data.creado,
  actualizado: data.actualizado,
});

// Servicio de Productos
export const productosService = {
  // Listar productos con filtros y paginación
  async getAll(filters?: ProductoFilters): Promise<PaginatedResponse<Producto>> {
    const params = new URLSearchParams();
    
    if (filters?.search) params.append('search', filters.search);
    if (filters?.categoria) params.append('categoria', filters.categoria.toString());
    if (filters?.precio_min) params.append('precio_min', filters.precio_min.toString());
    if (filters?.precio_max) params.append('precio_max', filters.precio_max.toString());
    if (filters?.en_oferta !== undefined) params.append('en_oferta', filters.en_oferta.toString());
    if (filters?.destacado !== undefined) params.append('destacado', filters.destacado.toString());
    if (filters?.activo !== undefined) params.append('activo', filters.activo.toString());
    if (filters?.stock_min) params.append('stock_min', filters.stock_min.toString());
    if (filters?.ordering) params.append('ordering', filters.ordering);
    if (filters?.page) params.append('page', filters.page.toString());
    if (filters?.page_size) params.append('page_size', filters.page_size.toString());

    const queryString = params.toString();
    const url = `/api/productos/${queryString ? `?${queryString}` : ''}`;
    
    const { data } = await apiRequest<PaginatedResponse<any>>(url);
    if (!data) throw new Error('No se pudo obtener los productos');
    
    return {
      ...data,
      results: data.results.map(fromDTO),
    };
  },

  // Obtener un producto por ID (incluye imágenes y variantes)
  async getById(id: number): Promise<Producto> {
    const { data } = await apiRequest<any>(`/api/productos/${id}/`);
    if (!data) throw new Error('No se pudo obtener el producto');
    return fromDTO(data);
  },

  // Crear un producto
  async create(data: ProductoFormData): Promise<Producto> {
    let body: FormData | string;
    
    // Si hay imagen, usar FormData
    if (data.imagen && data.imagen instanceof File) {
      const formData = new FormData();
      const dto = toDTO(data);
      
      // Agregar todos los campos al FormData
      Object.entries(dto).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          formData.append(key, value.toString());
        }
      });
      
      // Agregar la imagen
      formData.append('imagen', data.imagen);
      body = formData;
    } else {
      body = JSON.stringify(toDTO(data));
    }

    const { data: responseData } = await apiRequest<any>('/api/productos/', {
      method: 'POST',
      body,
    });
    if (!responseData) throw new Error('No se pudo crear el producto');
    return fromDTO(responseData);
  },

  // Actualizar un producto
  async update(id: number, data: Partial<ProductoFormData>): Promise<Producto> {
    let body: FormData | string;
    
    // Si hay imagen, usar FormData
    if (data.imagen && data.imagen instanceof File) {
      const formData = new FormData();
      const dto = toDTO(data as ProductoFormData);
      
      // Agregar todos los campos al FormData
      Object.entries(dto).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          formData.append(key, value.toString());
        }
      });
      
      // Agregar la imagen
      formData.append('imagen', data.imagen);
      body = formData;
    } else {
      body = JSON.stringify(toDTO(data as ProductoFormData));
    }

    const { data: responseData } = await apiRequest<any>(`/api/productos/${id}/`, {
      method: 'PATCH',
      body,
    });
    if (!responseData) throw new Error('No se pudo actualizar el producto');
    return fromDTO(responseData);
  },

  // Eliminar un producto
  async delete(id: number): Promise<void> {
    await apiRequest<void>(`/api/productos/${id}/`, {
      method: 'DELETE',
    });
  },

  // Obtener productos destacados
  async getDestacados(): Promise<Producto[]> {
    const { data } = await apiRequest<any[]>('/api/productos/destacados/');
    if (!data) throw new Error('No se pudo obtener los productos destacados');
    return data.map(fromDTO);
  },

  // Obtener productos en oferta
  async getOfertas(): Promise<Producto[]> {
    const { data } = await apiRequest<any[]>('/api/productos/ofertas/');
    if (!data) throw new Error('No se pudo obtener las ofertas');
    return data.map(fromDTO);
  },

  // Búsqueda inteligente de productos
  async buscar(query: string, page: number = 1, pageSize: number = 20): Promise<PaginatedResponse<Producto>> {
    const params = new URLSearchParams({
      q: query,
      page: page.toString(),
      page_size: pageSize.toString(),
    });
    
    const { data } = await apiRequest<any>(`/api/productos/buscar/?${params.toString()}`);
    if (!data) throw new Error('No se pudo realizar la búsqueda');
    
    return {
      count: data.count,
      next: data.next || null,
      previous: data.previous || null,
      results: data.results.map(fromDTO),
    };
  },

  // Obtener productos por categoría
  async porCategoria(categoriaId: number, filters?: ProductoFilters): Promise<PaginatedResponse<Producto>> {
    const params = new URLSearchParams({ categoria_id: categoriaId.toString() });
    
    if (filters?.page) params.append('page', filters.page.toString());
    if (filters?.page_size) params.append('page_size', filters.page_size.toString());
    if (filters?.ordering) params.append('ordering', filters.ordering);

    const { data } = await apiRequest<PaginatedResponse<any>>(
      `/api/productos/por_categoria/?${params.toString()}`
    );
    if (!data) throw new Error('No se pudo obtener los productos');
    
    return {
      ...data,
      results: data.results.map(fromDTO),
    };
  },
};

// Servicio de Categorías
export const categoriasService = {
  // Listar todas las categorías
  async getAll(): Promise<Categoria[]> {
    const { data } = await apiRequest<any>('/api/categorias/?page_size=100');
    if (!data) throw new Error('No se pudo obtener las categorías');
    
    // Manejar respuesta paginada o array directo
    const items = Array.isArray(data) ? data : (data.results || []);
    return items.map(categoriaFromDTO);
  },

  // Obtener una categoría por ID
  async getById(id: number): Promise<Categoria> {
    const { data } = await apiRequest<any>(`/api/categorias/${id}/`);
    if (!data) throw new Error('No se pudo obtener la categoría');
    return categoriaFromDTO(data);
  },

  // Crear una categoría
  async create(categoriaData: any): Promise<Categoria> {
    // Si hay imagen, usar FormData
    if (categoriaData.imagen instanceof File) {
      const formData = new FormData();
      formData.append('nombre', categoriaData.nombre);
      formData.append('descripcion', categoriaData.descripcion || '');
      formData.append('activa', categoriaData.activa !== undefined ? categoriaData.activa.toString() : 'true');
      formData.append('orden', categoriaData.orden?.toString() || '0');
      formData.append('imagen', categoriaData.imagen);

      const { data } = await apiRequest<any>('/api/categorias/', {
        method: 'POST',
        body: formData,
      });
      if (!data) throw new Error('No se pudo crear la categoría');
      return categoriaFromDTO(data);
    }
    
    // Sin imagen, usar JSON
    const payload = {
      nombre: categoriaData.nombre,
      descripcion: categoriaData.descripcion || '',
      activa: categoriaData.activa !== undefined ? categoriaData.activa : true,
      orden: categoriaData.orden || 0,
    };

    const { data } = await apiRequest<any>('/api/categorias/', {
      method: 'POST',
      body: JSON.stringify(payload),
      headers: {
        'Content-Type': 'application/json',
      },
    });
    if (!data) throw new Error('No se pudo crear la categoría');
    return categoriaFromDTO(data);
  },

  // Actualizar una categoría
  async update(id: number, categoriaData: Partial<any>): Promise<Categoria> {
    // Si hay imagen, usar FormData
    if (categoriaData.imagen instanceof File) {
      const formData = new FormData();
      
      if (categoriaData.nombre) formData.append('nombre', categoriaData.nombre);
      if (categoriaData.descripcion !== undefined) formData.append('descripcion', categoriaData.descripcion);
      if (categoriaData.activa !== undefined) formData.append('activa', categoriaData.activa.toString());
      if (categoriaData.orden !== undefined) formData.append('orden', categoriaData.orden.toString());
      formData.append('imagen', categoriaData.imagen);

      const { data } = await apiRequest<any>(`/api/categorias/${id}/`, {
        method: 'PATCH',
        body: formData,
      });
      if (!data) throw new Error('No se pudo actualizar la categoría');
      return categoriaFromDTO(data);
    }
    
    // Sin imagen, usar JSON
    const payload: any = {};
    if (categoriaData.nombre) payload.nombre = categoriaData.nombre;
    if (categoriaData.descripcion !== undefined) payload.descripcion = categoriaData.descripcion;
    if (categoriaData.activa !== undefined) payload.activa = categoriaData.activa;
    if (categoriaData.orden !== undefined) payload.orden = categoriaData.orden;

    const { data } = await apiRequest<any>(`/api/categorias/${id}/`, {
      method: 'PATCH',
      body: JSON.stringify(payload),
      headers: {
        'Content-Type': 'application/json',
      },
    });
    if (!data) throw new Error('No se pudo actualizar la categoría');
    return categoriaFromDTO(data);
  },

  // Eliminar una categoría
  async delete(id: number): Promise<void> {
    await apiRequest<void>(`/api/categorias/${id}/`, {
      method: 'DELETE',
    });
  },
};

// Servicio de Imágenes de Productos
export const productoImagenesService = {
  // Listar imágenes de un producto
  async getByProducto(productoId: number): Promise<ProductoImagen[]> {
    const { data } = await apiRequest<any[]>(`/api/imagenes/?producto=${productoId}`);
    if (!data) throw new Error('No se pudo obtener las imágenes');
    return data;
  },

  // Subir una imagen
  async upload(productoId: number, imagen: File, orden?: number, esPrincipal?: boolean): Promise<ProductoImagen> {
    const formData = new FormData();
    formData.append('producto', productoId.toString());
    formData.append('imagen', imagen);
    if (orden !== undefined) formData.append('orden', orden.toString());
    if (esPrincipal !== undefined) formData.append('es_principal', esPrincipal.toString());

    const { data } = await apiRequest<ProductoImagen>('/api/imagenes/', {
      method: 'POST',
      body: formData,
      headers: {}, // Dejar que el navegador establezca el Content-Type con boundary
    });
    if (!data) throw new Error('No se pudo subir la imagen');
    return data;
  },

  // Eliminar una imagen
  async delete(id: number): Promise<void> {
    await apiRequest<void>(`/api/imagenes/${id}/`, {
      method: 'DELETE',
    });
  },
};

// Servicio de Variantes de Productos
export const productoVariantesService = {
  // Listar variantes de un producto
  async getByProducto(productoId: number): Promise<ProductoVariante[]> {
    const { data } = await apiRequest<any[]>(`/api/variantes/?producto=${productoId}`);
    if (!data) throw new Error('No se pudo obtener las variantes');
    return data;
  },

  // Crear una variante
  async create(data: {
    producto: number;
    nombre: string;
    sku: string;
    precio_adicional?: number;
    stock: number;
  }): Promise<ProductoVariante> {
    const { data: responseData } = await apiRequest<ProductoVariante>('/api/variantes/', {
      method: 'POST',
      body: JSON.stringify(data),
    });
    if (!responseData) throw new Error('No se pudo crear la variante');
    return responseData;
  },

  // Actualizar una variante
  async update(id: number, data: Partial<ProductoVariante>): Promise<ProductoVariante> {
    const { data: responseData } = await apiRequest<ProductoVariante>(`/api/variantes/${id}/`, {
      method: 'PATCH',
      body: JSON.stringify(data),
    });
    if (!responseData) throw new Error('No se pudo actualizar la variante');
    return responseData;
  },

  // Eliminar una variante
  async delete(id: number): Promise<void> {
    await apiRequest<void>(`/api/variantes/${id}/`, {
      method: 'DELETE',
    });
  },
};

File: mobile/lib/services/productos_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import 'auth_service.dart';
import '../config/api_config.dart';

// Modelos
class Categoria {
  final int id;
  final String nombre;
  final String slug;
  final String? descripcion;
  final String? imagen;
  final bool activa;
  final int orden;
  final int productosCount;

  Categoria({
    required this.id,
    required this.nombre,
    required this.slug,
    this.descripcion,
    this.imagen,
    required this.activa,
    required this.orden,
    required this.productosCount,
  });

  factory Categoria.fromJson(Map<String, dynamic> json) {
    return Categoria(
      id: json['id'],
      nombre: json['nombre'],
      slug: json['slug'],
      descripcion: json['descripcion'],
      imagen: json['imagen'],
      activa: json['activa'] ?? true,
      orden: json['orden'] ?? 0,
      productosCount: json['productos_count'] ?? 0,
    );
  }
}

class ProductoImagen {
  final int id;
  final String imagen;
  final String? alt;
  final int orden;
  final bool esPrincipal;

  ProductoImagen({
    required this.id,
    required this.imagen,
    this.alt,
    required this.orden,
    required this.esPrincipal,
  });

  factory ProductoImagen.fromJson(Map<String, dynamic> json) {
    return ProductoImagen(
      id: json['id'],
      imagen: json['imagen'],
      alt: json['alt'],
      orden: json['orden'] ?? 0,
      esPrincipal: json['es_principal'] ?? false,
    );
  }
}

class Producto {
  final int id;
  final String nombre;
  final String slug;
  final String? descripcion;
  final String? descripcionCorta;
  final String? imagen;
  final int categoriaId;
  final String? categoriaNombre;
  final double precio;
  final bool enOferta;
  final double? precioOferta;
  final double precioFinal;
  final int? descuentoPorcentaje;
  final int stock;
  final int stockMinimo;
  final String? sku;
  final String? marca;
  final String? modelo;
  final bool activo;
  final bool destacado;
  final int vistas;
  final int ventas;
  final List<ProductoImagen> imagenes;

  Producto({
    required this.id,
    required this.nombre,
    required this.slug,
    this.descripcion,
    this.descripcionCorta,
    this.imagen,
    required this.categoriaId,
    this.categoriaNombre,
    required this.precio,
    required this.enOferta,
    this.precioOferta,
    required this.precioFinal,
    this.descuentoPorcentaje,
    required this.stock,
    required this.stockMinimo,
    this.sku,
    this.marca,
    this.modelo,
    required this.activo,
    required this.destacado,
    required this.vistas,
    required this.ventas,
    required this.imagenes,
  });

  factory Producto.fromJson(Map<String, dynamic> json) {
    return Producto(
      id: json['id'],
      nombre: json['nombre'],
      slug: json['slug'],
      descripcion: json['descripcion'],
      descripcionCorta: json['descripcion_corta'],
      imagen: json['imagen'],
      categoriaId: json['categoria'] is Map
          ? json['categoria']['id']
          : json['categoria'],
      categoriaNombre: json['categoria'] is Map
          ? json['categoria']['nombre']
          : null,
      precio: double.parse(json['precio'].toString()),
      enOferta: json['en_oferta'] ?? false,
      precioOferta: json['precio_oferta'] != null
          ? double.parse(json['precio_oferta'].toString())
          : null,
      precioFinal: double.parse(json['precio_final'].toString()),
      descuentoPorcentaje: json['descuento_porcentaje'],
      stock: json['stock'] ?? 0,
      stockMinimo: json['stock_minimo'] ?? 5,
      sku: json['sku'],
      marca: json['marca'],
      modelo: json['modelo'],
      activo: json['activo'] ?? true,
      destacado: json['destacado'] ?? false,
      vistas: json['vistas'] ?? 0,
      ventas: json['ventas'] ?? 0,
      imagenes: json['imagenes'] != null
          ? (json['imagenes'] as List)
                .map((img) => ProductoImagen.fromJson(img))
                .toList()
          : [],
    );
  }
}

class ProductosResponse {
  final int count;
  final String? next;
  final String? previous;
  final List<Producto> results;

  ProductosResponse({
    required this.count,
    this.next,
    this.previous,
    required this.results,
  });

  factory ProductosResponse.fromJson(Map<String, dynamic> json) {
    return ProductosResponse(
      count: json['count'],
      next: json['next'],
      previous: json['previous'],
      results: (json['results'] as List)
          .map((p) => Producto.fromJson(p))
          .toList(),
    );
  }
}

class ProductosService {
  final AuthService _authService = AuthService();

  String _getBaseUrl() {
    return ApiConfig.getBaseUrl();
  }

  // Exponer método para compartir
  String getBaseUrl() => _getBaseUrl();

  Future<Map<String, String>> _getHeaders() async {
    final token = await _authService.getToken();
    return {
      'Content-Type': 'application/json',
      if (token != null) 'Authorization': 'Bearer $token',
    };
  }

  // Listar productos con filtros
  Future<ProductosResponse> getProductos({
    int page = 1,
    String? search,
    String? categoriaId,
    double? precioMin,
    double? precioMax,
    bool? enOferta,
    bool? destacado,
    String? ordenamiento,
  }) async {
    try {
      final baseUrl = _getBaseUrl();
      final headers = await _getHeaders();

      // Construir query parameters
      final queryParams = <String, String>{
        'page': page.toString(),
        if (search != null && search.isNotEmpty) 'search': search,
        if (categoriaId != null) 'categoria': categoriaId,
        if (precioMin != null) 'precio_min': precioMin.toString(),
        if (precioMax != null) 'precio_max': precioMax.toString(),
        if (enOferta != null) 'en_oferta': enOferta.toString(),
        if (destacado != null) 'destacado': destacado.toString(),
        if (ordenamiento != null) 'ordering': ordenamiento,
      };

      final uri = Uri.parse(
        '$baseUrl/api/productos/',
      ).replace(queryParameters: queryParams);

      final response = await http.get(uri, headers: headers);

      if (response.statusCode == 200) {
        final decodedBody = utf8.decode(response.bodyBytes);
        final data = json.decode(decodedBody);
        return ProductosResponse.fromJson(data);
      } else {
        throw Exception('Error al cargar productos: ${response.statusCode}');
      }
    } catch (e) {
      print('Error en getProductos: $e');
      rethrow;
    }
  }

  // Obtener un producto por ID
  Future<Producto> getProducto(int id) async {
    try {
      final baseUrl = _getBaseUrl();
      final headers = await _getHeaders();

      final response = await http.get(
        Uri.parse('$baseUrl/api/productos/$id/'),
        headers: headers,
      );

      if (response.statusCode == 200) {
        final decodedBody = utf8.decode(response.bodyBytes);
        final data = json.decode(decodedBody);
        return Producto.fromJson(data);
      } else {
        throw Exception('Error al cargar producto: ${response.statusCode}');
      }
    } catch (e) {
      print('Error en getProducto: $e');
      rethrow;
    }
  }

  // Obtener productos destacados
  Future<List<Producto>> getProductosDestacados({int limit = 10}) async {
    try {
      final response = await getProductos(
        destacado: true,
        ordenamiento: '-vistas',
      );
      return response.results.take(limit).toList();
    } catch (e) {
      print('Error en getProductosDestacados: $e');
      rethrow;
    }
  }

  // Obtener productos en oferta
  Future<List<Producto>> getProductosEnOferta({int limit = 10}) async {
    try {
      final response = await getProductos(
        enOferta: true,
        ordenamiento: '-descuento_porcentaje',
      );
      return response.results.take(limit).toList();
    } catch (e) {
      print('Error en getProductosEnOferta: $e');
      rethrow;
    }
  }

  // Obtener categorías
  Future<List<Categoria>> getCategorias() async {
    try {
      final baseUrl = _getBaseUrl();
      final headers = await _getHeaders();

      final response = await http.get(
        Uri.parse('$baseUrl/api/categorias/'),
        headers: headers,
      );

      if (response.statusCode == 200) {
        final decodedBody = utf8.decode(response.bodyBytes);
        final data = json.decode(decodedBody);

        // El backend devuelve con paginación, extraer results
        final List resultsList = data is Map ? (data['results'] ?? []) : data;
        return resultsList.map((cat) => Categoria.fromJson(cat)).toList();
      } else {
        throw Exception('Error al cargar categorías: ${response.statusCode}');
      }
    } catch (e) {
      print('Error en getCategorias: $e');
      rethrow;
    }
  }

  // Buscar productos
  Future<List<Producto>> buscarProductos(String query) async {
    try {
      final response = await getProductos(search: query);
      return response.results;
    } catch (e) {
      print('Error en buscarProductos: $e');
      rethrow;
    }
  }

  // Buscar producto por código de barras
  Future<Producto?> buscarPorCodigoBarras(String codigo) async {
    try {
      final response = await getProductos(search: codigo);
      if (response.results.isNotEmpty) {
        // Buscar coincidencia exacta en SKU o código de barras
        final producto = response.results.firstWhere(
          (p) => p.sku?.toLowerCase() == codigo.toLowerCase(),
          orElse: () => response.results.first,
        );
        return producto;
      }
      return null;
    } catch (e) {
      print('Error en buscarPorCodigoBarras: $e');
      return null;
    }
  }

  // Historial de búsquedas
  static const String _historialKey = 'historial_busquedas';
  static const int _maxHistorial = 10;

  Future<void> guardarBusqueda(String query) async {
    if (query.trim().isEmpty) return;

    try {
      final prefs = await SharedPreferences.getInstance();
      final historial = await getBusquedasRecientes();
      
      // Eliminar si ya existe
      historial.remove(query.trim());
      
      // Agregar al inicio
      historial.insert(0, query.trim());
      
      // Limitar tamaño
      if (historial.length > _maxHistorial) {
        historial.removeRange(_maxHistorial, historial.length);
      }
      
      await prefs.setStringList(_historialKey, historial);
    } catch (e) {
      print('Error guardando búsqueda: $e');
    }
  }

  Future<List<String>> getBusquedasRecientes() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      return prefs.getStringList(_historialKey) ?? [];
    } catch (e) {
      print('Error obteniendo historial: $e');
      return [];
    }
  }

  Future<void> limpiarHistorial() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.remove(_historialKey);
    } catch (e) {
      print('Error limpiando historial: $e');
    }
  }

  // Filtros guardados
  static const String _filtrosKey = 'filtros_guardados';
  static const int _maxFiltros = 5;

  Future<void> guardarFiltro({
    String? categoriaId,
    double? precioMin,
    double? precioMax,
    bool? enOferta,
    bool? destacado,
    String? ordenamiento,
  }) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final filtrosGuardados = await getFiltrosGuardados();

      final nuevoFiltro = {
        'categoriaId': categoriaId,
        'precioMin': precioMin,
        'precioMax': precioMax,
        'enOferta': enOferta,
        'destacado': destacado,
        'ordenamiento': ordenamiento,
        'nombre': _generarNombreFiltro(
          categoriaId: categoriaId,
          precioMin: precioMin,
          precioMax: precioMax,
          enOferta: enOferta,
          destacado: destacado,
        ),
        'fecha': DateTime.now().toIso8601String(),
      };

      // Eliminar si ya existe un filtro idéntico
      filtrosGuardados.removeWhere((f) => _sonFiltrosIguales(f, nuevoFiltro));

      // Agregar al inicio
      filtrosGuardados.insert(0, nuevoFiltro);

      // Limitar tamaño
      if (filtrosGuardados.length > _maxFiltros) {
        filtrosGuardados.removeRange(_maxFiltros, filtrosGuardados.length);
      }

      final filtrosJson = jsonEncode(filtrosGuardados);
      await prefs.setString(_filtrosKey, filtrosJson);
    } catch (e) {
      print('Error guardando filtro: $e');
    }
  }

  Future<List<Map<String, dynamic>>> getFiltrosGuardados() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final filtrosJson = prefs.getString(_filtrosKey);
      if (filtrosJson == null) return [];

      final List<dynamic> filtrosList = jsonDecode(filtrosJson);
      return filtrosList.cast<Map<String, dynamic>>();
    } catch (e) {
      print('Error obteniendo filtros guardados: $e');
      return [];
    }
  }

  Future<void> eliminarFiltro(int index) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final filtrosGuardados = await getFiltrosGuardados();
      if (index >= 0 && index < filtrosGuardados.length) {
        filtrosGuardados.removeAt(index);
        final filtrosJson = jsonEncode(filtrosGuardados);
        await prefs.setString(_filtrosKey, filtrosJson);
      }
    } catch (e) {
      print('Error eliminando filtro: $e');
    }
  }

  Future<void> limpiarFiltrosGuardados() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.remove(_filtrosKey);
    } catch (e) {
      print('Error limpiando filtros guardados: $e');
    }
  }

  String _generarNombreFiltro({
    String? categoriaId,
    double? precioMin,
    double? precioMax,
    bool? enOferta,
    bool? destacado,
  }) {
    final partes = <String>[];

    if (categoriaId != null) {
      final categoria = _categoriasCache.firstWhere(
        (c) => c.id.toString() == categoriaId,
        orElse: () => Categoria(
          id: int.parse(categoriaId),
          nombre: 'Categoría',
          slug: '',
          activa: true,
          orden: 0,
          productosCount: 0,
        ),
      );
      partes.add(categoria.nombre);
    }

    if (precioMin != null || precioMax != null) {
      if (precioMin != null && precioMax != null) {
        partes.add('Bs. ${precioMin.toStringAsFixed(0)}-${precioMax.toStringAsFixed(0)}');
      } else if (precioMin != null) {
        partes.add('Desde Bs. ${precioMin.toStringAsFixed(0)}');
      } else {
        partes.add('Hasta Bs. ${precioMax!.toStringAsFixed(0)}');
      }
    }

    if (enOferta == true) partes.add('Ofertas');
    if (destacado == true) partes.add('Destacados');

    return partes.isEmpty ? 'Filtro personalizado' : partes.join(' • ');
  }

  bool _sonFiltrosIguales(Map<String, dynamic> f1, Map<String, dynamic> f2) {
    return f1['categoriaId'] == f2['categoriaId'] &&
        f1['precioMin'] == f2['precioMin'] &&
        f1['precioMax'] == f2['precioMax'] &&
        f1['enOferta'] == f2['enOferta'] &&
        f1['destacado'] == f2['destacado'] &&
        f1['ordenamiento'] == f2['ordenamiento'];
  }

  // Cache temporal de categorías para generar nombres
  static List<Categoria> _categoriasCache = [];
  static void setCategoriasCache(List<Categoria> categorias) {
    _categoriasCache = categorias;
  }
}
